File: backend/src/main/java/github/lms/lemuel/xr/asset/application/AssetManifestSeeder.java
package github.lms.lemuel.xr.asset.application;

import com.fasterxml.jackson.databind.ObjectMapper;
import github.lms.lemuel.xr.asset.adapter.out.persistence.AssetManifestJpaEntity;
import github.lms.lemuel.xr.asset.adapter.out.persistence.AssetManifestRepository;
import jakarta.annotation.PostConstruct;
import java.io.InputStream;
import java.time.LocalDateTime;
import java.util.Map;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.io.Resource;
import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
import org.springframework.core.io.support.ResourcePatternResolver;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

/**
 * classpath:manifests/{mission}/{device}/scene{N}-v{version}.json 을 스캔해
 * asset_manifests 테이블에 upsert. 부팅 1회.
 *
 * <p>(mission_id, scene_number, device_type, version) 이미 있으면 스킵 — 재배포 안전.
 * 새 버전이면 신규 row INSERT (V10 의 superseded_by 흐름은 운영 도구에서 처리).</p>
 */
@Component
@Slf4j
@RequiredArgsConstructor
public class AssetManifestSeeder {

    private final AssetManifestRepository repo;
    private final ObjectMapper json;

    @PostConstruct
    @Transactional
    public void seed() {
        ResourcePatternResolver scanner = new PathMatchingResourcePatternResolver();
        Resource[] files;
        try {
            files = scanner.getResources("classpath*:manifests/**/*.json");
        } catch (Exception e) {
            log.warn("manifest 스캔 실패: {}", e.getMessage());
            return;
        }

        int inserted = 0;
        int skipped = 0;
        for (Resource r : files) {
            try (InputStream in = r.getInputStream()) {
                @SuppressWarnings("unchecked")
                Map<String, Object> doc = json.readValue(in, Map.class);

                String missionId = (String) doc.get("mission_id");
                Short sceneNumber = doc.get("scene_number") == null ? null : ((Number) doc.get("scene_number")).shortValue();
                String deviceType = (String) doc.get("device_type");
                String version = (String) doc.get("version");

                if (missionId == null || deviceType == null || version == null) {
                    log.warn("필수 필드 누락 — 건너뜀: {}", r.getFilename());
                    continue;
                }

                if (repo.existsByMissionIdAndSceneNumberAndDeviceTypeAndVersion(
                        missionId, sceneNumber, deviceType, version)) {
                    skipped++;
                    continue;
                }

                AssetManifestJpaEntity row = new AssetManifestJpaEntity();
                row.setId(UUID.randomUUID());
                row.setMissionId(missionId);
                row.setSceneNumber(sceneNumber);
                row.setDeviceType(deviceType);
                row.setVersion(version);
                row.setCapabilitiesMin(asMap(doc.get("capabilities_min")));
                row.setManifest(asMap(doc.get("manifest")));
                row.setAudioLocale((String) doc.get("audio_locale"));
                row.setTotalSizeBytes(doc.get("total_size_bytes") == null
                        ? null : ((Number) doc.get("total_size_bytes")).longValue());
                row.setCdnBaseUrl((String) doc.get("cdn_base_url"));
                row.setIsActive(true);
                row.setCreatedAt(LocalDateTime.now());

                repo.save(row);
                inserted++;
            } catch (Exception e) {
                log.error("manifest 로드 실패 — {}: {}", r.getFilename(), e.getMessage(), e);
            }
        }
        log.info("asset_manifests seed 완료: inserted={}, skipped={} (총 {} 파일 스캔)",
                inserted, skipped, files.length);
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> asMap(Object v) {
        return v == null ? null : (Map<String, Object>) v;
    }
}

File: backend/src/main/kotlin/github/lms/lemuel/xr/asset/application/AssetManifestSeeder.kt
package github.lms.lemuel.xr.asset.application

import github.lms.lemuel.xr.asset.application.port.out.AssetManifestPort
import github.lms.lemuel.xr.asset.application.seed.ManifestDocument
import github.lms.lemuel.xr.asset.application.seed.ManifestParser
import github.lms.lemuel.xr.asset.application.seed.ManifestScanner
import github.lms.lemuel.xr.asset.application.seed.ManifestValidator
import github.lms.lemuel.xr.asset.domain.AssetManifest
import jakarta.annotation.PostConstruct
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.UUID

/**
 * classpath:manifests/{mission}/{device}/scene{N}-v{version}.json 을 스캔해
 * asset_manifests 테이블에 upsert. 부팅 1회.
 *
 * (mission_id, scene_number, device_type, version) 이미 있으면 스킵 — 재배포 안전.
 * 새 버전이면 신규 row INSERT (V10 의 superseded_by 흐름은 운영 도구에서 처리).
 *
 * SRP — 스캔([ManifestScanner])·파싱([ManifestParser])·검증
 * ([ManifestValidator]) 은 각각 별도 컴포넌트로 분리했고, 여기서는 그 셋을
 * 순서대로 엮어(orchestrate) 도메인 [AssetManifest] 구성·영속화만 담당한다.
 * 엔티티 매핑은 포트 어댑터가 전담하므로 시더는 JPA 를 알지 못한다.
 */
@Component
class AssetManifestSeeder(
    private val manifests: AssetManifestPort,
    private val scanner: ManifestScanner,
    private val parser: ManifestParser,
    private val validator: ManifestValidator,
) {

    @PostConstruct
    @Transactional
    fun seed() {
        val files = scanner.scan()

        var inserted = 0
        var skipped = 0
        for (r in files) {
            try {
                val doc = parser.parse(r)

                if (!validator.isValid(doc)) {
                    log.warn("필수 필드 누락 — 건너뜀: {}", r.filename)
                    continue
                }

                if (manifests.existsByCoordinates(
                        doc.missionId!!, doc.sceneNumber, doc.deviceType!!, doc.version!!,
                    )
                ) {
                    skipped++
                    continue
                }

                manifests.save(toDomain(doc))
                inserted++
            } catch (e: Exception) {
                log.error("manifest 로드 실패 — {}: {}", r.filename, e.message, e)
            }
        }
        log.info(
            "asset_manifests seed 완료: inserted={}, skipped={} (총 {} 파일 스캔)",
            inserted, skipped, files.size,
        )
    }

    /** 파싱된 [ManifestDocument] 를 신규 도메인 [AssetManifest] 로 (id·createdAt 생성, active=true). */
    private fun toDomain(doc: ManifestDocument): AssetManifest =
        AssetManifest(
            id = UUID.randomUUID(),
            missionId = doc.missionId!!,
            sceneNumber = doc.sceneNumber,
            deviceType = doc.deviceType!!,
            capabilitiesMin = doc.capabilitiesMin,
            version = doc.version!!,
            manifest = doc.manifest ?: emptyMap(),
            audioLocale = doc.audioLocale,
            totalSizeBytes = doc.totalSizeBytes,
            cdnBaseUrl = doc.cdnBaseUrl,
            active = true,
            createdAt = LocalDateTime.now(),
            supersededAt = null,
        )

    companion object {
        private val log = LoggerFactory.getLogger(AssetManifestSeeder::class.java)
    }
}
